extensions/interactions/base_test.py

```diff
@@ -246,6 +247,142 @@ def test_interaction_rule_descriptions_in_dict(self):
+        # Contents of the file html_field_types_to_rule_specs.json will be
+        # verified against this file.
+        rule_descriptions_dict = json.loads(
+            utils.get_file_contents(feconf.RULES_DESCRIPTIONS_FILE_PATH))
+
+
+        # In the following part, we generate the html_field_types_to_rule_specs
+        # dict based on the values in the rule_descriptions.json file.
+        generated_html_field_types_dict = {}
+        html_types = (
+            [{
+                'type': 'DragAndDropHtmlString',
+                'format': 'string'
+            }, {
+                'type': 'ListOfSetsOfHtmlStrings',
+                'format': 'listOfSets'
+            }, {
+                'type': 'SetOfHtmlString',
+                'format': 'set'
+            }])
+
+        for interaction_id, interaction_rules in (
+                rule_descriptions_dict.items()):
+            for rule_type, rule_description in interaction_rules.items():
+                description = rule_description['description']
+                for html_type in html_types:
+                    if html_type['type'] in description:
+                        if (
+                                html_type['type'] not in
+                                generated_html_field_types_dict):
+                            generated_html_field_types_dict[
+                                html_type['type']] = {}
+                        generated_html_field_types_dict[
+                            html_type['type']]['interactionId'] = (
+                                interaction_id)
+                        generated_html_field_types_dict[html_type['type']][
+                            'format'] = html_type['format']
+                        if (
+                                'ruleTypes' not in
+                                generated_html_field_types_dict[
+                                    html_type['type']]):
+                            generated_html_field_types_dict[html_type['type']][
+                                'ruleTypes'] = {}
+                        generated_html_field_types_dict[html_type['type']][
+                            'ruleTypes'][rule_type] = {}
+                        generated_html_field_types_dict[html_type['type']][
+                            'ruleTypes'][rule_type]['htmlInputVariables'] = (
+                                ['x'])
+                        if '{{y|%s}}' % html_type['type'] in description:
+                            generated_html_field_types_dict[html_type['type']][
+                                'ruleTypes'][rule_type][
+                                    'htmlInputVariables'].append('y')
+
+        # Assert that the generated dict matches dict in the stored file.
+        self.assertEqual(
+            html_field_types_to_rule_specs_dict,
+            generated_html_field_types_dict)
+
+        # We verify the structure of the new file by generating a schema from
```

Comment:
No, this is wrong. You have already established that the generated dict is equal to the desired dict. So you are basically taking a dict, computing its schema, and then asserting that the dict matches the schema, which is not very useful.

The schema to match against should be **hardcoded** here. You can either use custom validation for the inner dicts (keyed by rule type) or you can squash each inner dict down to:

```
{
    ruleName: "IsEqualToOrdering",
    htmlInputVariables: ["x"]
},
```
and make ruleTypes a list of these.

Reply:
@seanlip So is the schema validation still needed, because we have already generated a dict and verified it(the schema is checked indirectly while generating the dict itself, so checking the schema again is a bit redundant right?)

Reply:
Oh, I see what you mean. Yes, that is fair -- the generation and the equality check does account for the entire dict. So I think it's OK to omit the schema validation. Thanks for pointing that out.

---

The second part is now obsolete, but I can explain -- in tests, you usually have a pattern where the expected value is hardcoded and the observed value is computed, and the two are compared. That's what I was referring to here.

Reply:
done